### PR TITLE
Prevent process polyfill to be included in builds because of process …

### PR DESCRIPTION
**Pull request by Andarist**
Related - https://github.com/reactjs/redux/issues/2244

By guarding `process` like this its polyfill is included by default when bundled with webpack. It should be highly desirable to avoid such side effect. 

I understand that this PR is not complete, but as your build setup is way more complex than what is used in most projects I would need some guidelines how this should be proceeded further. Maybe introducing `__TEST__` in similar manner how u introduce `__DEV__`?

Also this should land in `15.x` branch too and aint sure how this should be tackled here.

cc @gaearon 

Also maybe that could have been tackled on webpack's side as most of the people dont realise this is a problem and guarding against `process` seems like a right way to do things in the code. Dunno if we can safely predict what was the user intention though. cc @TheLarkInn @sokra

### CODE REVIEW

File: src/isomorphic/children/flattenChildren.js [hunk 1]
```diff
@@ -18,11 +18,7 @@ var warning = require('fbjs/lib/warning');
 
 var ReactComponentTreeHook;
 
-if (
-  typeof process !== 'undefined' &&
-  process.env &&
-  process.env.NODE_ENV === 'test'
-) {
```
Comment (gaearon) [inline src/isomorphic/children/flattenChildren.js]: Please just change this require to be unconditional. It is no longer necessary to do this trickery now. (The original reason for this was related to some Jest behavior that's irrelevant with flat bundles.)

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js [hunk 2]
```diff
@@ -22,11 +22,7 @@ var warning = require('fbjs/lib/warning');
 
 var ReactComponentTreeHook;
 
-if (
-  typeof process !== 'undefined' &&
-  process.env &&
-  process.env.NODE_ENV === 'test'
-) {
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Let's not touch this file. It's gone in next version of React.

File: src/renderers/shared/stack/reconciler/flattenStackChildren.js [hunk 3]
```diff
@@ -18,11 +18,7 @@ var warning = require('fbjs/lib/warning');
 
 var ReactComponentTreeHook;
 
-if (
-  typeof process !== 'undefined' &&
-  process.env &&
-  process.env.NODE_ENV === 'test'
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/flattenStackChildren.js]: Same here, let's leave this one as is because it's a goner.

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js (hunk 2 above)
Comment (Andarist) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Should i revert this change and the latter? Would be cool to fix this whole thing in `15.x`

File: src/isomorphic/children/flattenChildren.js (hunk 1 above)
Comment (Andarist) [inline src/isomorphic/children/flattenChildren.js]: should i just `var ReactComponentTreeHook = require('ReactComponentTreeHook')` at the top of the file? 

This `ReactComponentTreeHook` is later conditionally required in `__DEV__` block. What about it then? Should something be adjusted?

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js (hunk 2 above)
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Yes, please revert. I don’t see us fixing this in 15 given that the hack was added to work around an accidental breaking change in 15.2.1: https://github.com/facebook/react/issues/7240. Let’s wait for 16 here please.

File: src/isomorphic/children/flattenChildren.js (hunk 1 above)
Comment (gaearon) [inline src/isomorphic/children/flattenChildren.js]: >should i just var ReactComponentTreeHook = require('ReactComponentTreeHook') at the top of the file?

Yes.

>This ReactComponentTreeHook is later conditionally required in __DEV__ block.

You can remove that. It’s unnecessary now.

**Comment by Andarist** [comment]
@gaearon ok, so i've amended the commit according to your suggestions

**Comment by gaearon** [comment]
Can you run `yarn build -- core --type=UMD` and verify code from `ReactComponentTreeHook` doesn't end up in `react.production.min.js`?

**Comment by gaearon** [comment]
This looks good to me, thanks!

**Comment by Andarist** [comment]
@gaearon sorry for not confirming sooner, ive tried to figure this out sooner, but the mentioned file [`src/isomorphic/children/flattenChildren.js`](https://github.com/facebook/react/blob/50d6f9656d9d67643e1ff66300a41997f1753325/src/isomorphic/children/flattenChildren.js) is in fact not a part of most build types and I had to return to the issue way later than ive expected due to lack of time

Had to realise its used only for `yarn build -- core --type=FB` and now I can confirm that the goal is satisfied ofc. `ReactComponentTreeHook` is gone for production build
